boot/bootstate20.go

```diff
@@ -21,228 +21,176 @@ package boot
-	if err != nil {
-		return nil, nil, false, fmt.Errorf("cannot identify try kernel snap with bootloader %s: %v", ebl.Name(), err)
-	}
+	ks20.ebl = ebl
 
-	if tryKernelExists {
-		tryBootSnap = tryKernel
-	}
-
-	m, err := bl.GetBootVars("kernel_status")
+	// also get the kernel_status
+	m, err := ebl.GetBootVars("kernel_status")
 	if err != nil {
-		return nil, nil, false, fmt.Errorf("cannot read boot variables with bootloader %s: %v", ebl.Name(), err)
+		return err
 	}
 
-	trying = m["kernel_status"] == "trying"
+	ks20.kernelStatus = m["kernel_status"]
 
-	return bootSnap, tryBootSnap, trying, nil
+	return nil
 }
 
-func (s20 *bootState20Kernel) markSuccessful(update bootStateUpdate) (bootStateUpdate, error) {
-	var u20 *bootStateUpdate20
-	var err error
-	u20, err = newBootStateUpdate20(update, snap.TypeKernel)
+func (ks20 *bootState20Kernel) revisions() (snap.PlaceInfo, snap.PlaceInfo, bool, error) {
+	var bootSn, tryBootSn snap.PlaceInfo
+	var trying bool
+	err := ks20.loadBootloader()
 	if err != nil {
-		return nil, err
+		return nil, nil, false, err
 	}
 
-	// kernel_status goes from "" -> "try" -> "trying" (set by the boot script) -> ""
-	// so if we are not in "trying" mode, nothing to do here
-	if u20.env["kernel_status"] != "trying" {
-		return u20, nil
+	// get the kernel for this bootloader
+	bootSn, err = ks20.ebl.Kernel()
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("cannot identify kernel snap with bootloader %s: %v", ks20.ebl.Name(), err)
 	}
 
-	// if kernel_status is trying, then we need to figure out what kernel we did
-	// boot from
-	sn, exists, err := u20.ebl.TryKernel()
+	tryKernel, tryKernelExists, err := ks20.ebl.TryKernel()
 	if err != nil {
-		return nil, err
+		return nil, nil, false, fmt.Errorf("cannot identify try kernel snap with bootloader %s: %v", ks20.ebl.Name(), err)
 	}
 
-	if !exists {
-		// somehow we booted with kernel_status=trying, but no try-kernel
-		// symlink :-/
-		return nil, fmt.Errorf("cannot mark boot successful: kernel_status is \"trying\", but no try-kernel enabled")
+	if tryKernelExists {
+		tryBootSn = tryKernel
+	}
+
+	m, err := ks20.ebl.GetBootVars("kernel_status")
+	if err != nil {
+		return nil, nil, false, fmt.Errorf("cannot read boot variables with bootloader %s: %v", ks20.ebl.Name(), err)
 	}
+	kernelStatus := m["kernel_status"]
 
-	// set the snap we tried to the snap from the try-kernel symlink
-	u20.snapTried = sn
+	trying = (kernelStatus == "trying")
 
-	// update the boot var
-	u20.toCommit["kernel_status"] = ""
+	if kernelStatus != "" && tryBootSn == nil {
+		// we don't have a try kernel snap, but kernel_status is "", so we
+		// should tell markSuccessful to clean that up
+		// note we don't set kernelStatus, since revisions() could be used with
+		// setNext() and this would mess that up
+		logger.Noticef("kernel_status is %q, but there is no try kernel snap on bootloader %s, resetting to \"\"", kernelStatus, ks20.ebl.Name())
+		ks20.markSuccessCleanStatus = true
```

Comment:
it's a bit odd to do that here,  it should be done one level up I think, returning status string should help with that